packages/test/test-utils/src/loaderContainerTracker.ts

```diff
@@ -463,45 +465,57 @@ export class LoaderContainerTracker implements IOpProcessingController {
+                        if (contents.contents?.address !== undefined) {
+                            address += `/${contents.contents.address}`;
+                            contents = contents.contents.contents;
+                        } else if (contents.content?.address !== undefined) {
+                            address += `/${contents.content.address}`;
+                            contents = contents.content.contents;
+                        } else {
+                            break;
+                        }
+                    }
+                    if (address) {
+                        return `${address} ${JSON.stringify(contents)}`;
+                    }
+                    return JSON.stringify(contents);
+                } catch (e) {
+                    return `${e.message}: ${e.stack}`;
                 }
-                if (address) {
-                    return `${address} ${JSON.stringify(contents)}`;
-                }
-                return JSON.stringify(contents);
             };
             debugOp(`${index}: ADD: clientId: ${(container as Container).clientId}`);
             container.deltaManager.outbound.on("op", (messages) => {
                 for (const msg of messages) {
                     debugOp(`${index}: OUT:          `
                         + `cli: ${msg.clientSequenceNumber.toString().padStart(3)} `
-                        + `         ${msg.type} ${getContentsString(msg.type, msg.contents)}`);
+                        + `rsq: ${msg.referenceSequenceNumber.toString().padStart(3)} `
+                        + `${msg.type} ${getContentsString(msg.type, msg.contents)}`);
                 }
             });
-            container.deltaManager.inbound.on("push", (msg) => {
-                const clientSeq = msg.clientId === (container as Container).clientId ?
-                    `cli: ${msg.clientSequenceNumber.toString().padStart(3)}` : "        ";
-                debugOp(`${index}: IN : seq: ${msg.sequenceNumber.toString().padStart(3)} `
-                    + `${clientSeq} min: ${msg.minimumSequenceNumber.toString().padStart(3)} `
-                    + `${msg.type} ${getContentsString(msg.type, msg.contents)}`);
-            });
-
+            const getInboundHandler = (type: string) => {
+                return (msg: ISequencedDocumentMessage) => {
+                    const clientSeq = msg.clientId === (container as Container).clientId ?
+                        `cli: ${msg.clientSequenceNumber.toString().padStart(3)}` : "        ";
+                    debugOp(`${index}: ${type}: seq: ${msg.sequenceNumber.toString().padStart(3)} `
+                        + `${clientSeq} min: ${msg.minimumSequenceNumber.toString().padStart(3)} `
+                        + `${msg.type} ${getContentsString(msg.type, msg.contents)}`);
+                };
+            };
+            container.deltaManager.inbound.on("push", getInboundHandler("IN "));
```

Comment:
Why do we need to log both "push" and "op" events for the same op? What additional information do we out of it?

Reply:
`push` happens when we received the message on the wire and it get queued up in the DeltaQueue.
`op` happens when the message leaves the DeltaQueue and delivered to the runtime.

Reply:
Yeah, but I was wondering why do we need to log both since it would be the same op :). We don't do this for outbound ops.

Reply:
When I was debugging real life test, I want to make sure when things happen. When `processOutgoing` was called, we send the op and wait until those ops comes back in the queue.  But I don't want those op to be processed by the runtime yet.    It allow for more fine grain logging of what is going on.